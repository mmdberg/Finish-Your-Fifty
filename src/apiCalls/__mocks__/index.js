export const fetchUsers = jest.fn().mockImplementation(() => Promise.resolve(
  [{
    userName: 'Pizza',
    email: 'pizza@pizza',
    password: 'p'
  }]
));

export const addUser = jest.fn().mockImplementation(() => Promise.resolve(1));

export const fetchOneUser = jest.fn()
  .mockImplementationOnce(() => Promise.rejects())
  .mockImplementation(() => Promise.resolve({ 
    id: 7,
    userName: 'Pizza',
    email: 'pizza@pizza',
    password: 'p'
}));