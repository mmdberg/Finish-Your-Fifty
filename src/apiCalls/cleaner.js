export const raceCleaner = (races) => {
  const cleanRaces = races.map(race => {
    return {
      city: race.city,
      state: race.state,
      date: dateCleaner(race.start_date_local),
      raceName: race.name,
      id: race.id
    };
  });
  return cleanRaces;
};

export const dateCleaner = (date) => {
  const dateOnly = date.split('').splice(0, 10);
  const yearOnly = dateOnly.splice(0, 5).splice(0, 4);
  const cleanDate = [...date.split('').splice(5, 5), '-', ...yearOnly].join('');
  return cleanDate;
};