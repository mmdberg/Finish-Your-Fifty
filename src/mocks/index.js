export const mockRace = {
  venue: 'Atlas Park',
  city: 'San Diego',
  event: 'Spooky 5k',
  date: '04-05-2018'
};

export const mockApiResult = {
  place: {
    placeName: 'Atlas Park',
    cityName: 'San Diego'
  },
  activityStartDate: '2018-04-05T08:00:00',
  organization: {
    organizationName: 'Spooky 5k'
  }
};

export const mockUser = {
  name: 'Pizza',
  email: 'pizza@pizza',
  password: 'p'
};