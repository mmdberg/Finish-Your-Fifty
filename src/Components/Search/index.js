import React, { Component } from 'react';
import * as api from '../../apiCalls';
import './styles.css';
import ReactTable from 'react-table';
import 'react-table/react-table.css';


            // <table>
            //   <tr>
            //     <th>Event</th>
            //     <th>City</th>
            //     <th>Date</th>
            //     <th>State</th>
            //   </tr>
            //   {this.state.results}
            // </table>

class Search extends Component {
  constructor(props) {
    super(props);
    this.state = {
      state: '',
      year: '',
      results: [],
      error: ''
    };
  }

  handleChange = (event) => {
    const {name, value} = event.target;
    this.setState({
      [name]: value.toUpperCase()
    });
  }

  handleSubmit = async (event) => {
    event.preventDefault();
    const results = await api.fetchRaces(this.state.year);
    const racesByState = results.filter(race => 
      race.state === this.state.state
    )
    if (!racesByState.length) {
      this.setState({
        error: 'No races match your search criteria. Try again.'
      });
    } else {
      this.setState({
        results: racesByState
      });
    }
  }

  clearSearch = () => {
    this.setState({
      state: '',
      year: '',
      results: [],
      error: ''
    }); 
  }

  render() {
    return (
      <div>
        <h2>Search for races by state</h2>
        <form>
          <input type='text'
            name='state'
            placeholder='State (ex. CA)'
            value={this.state.state}
            onChange={this.handleChange}
            maxLength='2'
          />
          <input type='text'
          name='year'
          placeholder='Year'
          value={this.state.year}
          onChange={this.handleChange}
          maxLength='4'/>
          <button onClick={this.handleSubmit}>Search</button>
          <button onClick={this.clearSearch}>Clear</button>
        </form>
        {
          this.state.results.length ?
            <ReactTable
              data={this.state.results}
              columns={[
                {
                  Header: 'Event',
                  accessor: 'event'
                },
                {
                  Header: 'City',
                  accessor: 'city'
                },
                {
                  Header: 'Date',
                  accessor: 'date'
                },
                {
                  Header: 'State',
                  accessor: 'state'
                }
              ]} />
            : this.state.error
        }

      </div>
    );
  }

}

export default Search;