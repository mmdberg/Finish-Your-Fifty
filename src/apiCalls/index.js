import { apiKey } from '../private/apiKey';
import { raceCleaner } from './cleaner'

export const fetchRaces = async (state) => { 
  try {
    const response 
      = await fetch(`/v2/search/?query=running&category=event&state=${state}&api_key=${apiKey}`,
        {
          type: 'GET',
          credentials: 'omit'
        });
    const raceInfo = await response.json();
    return raceCleaner(raceInfo.results);
  } catch (error) {
    console.log('error:', error);
  }
};



export const fetchUsers = async () => {
  try {
    const response = await fetch('http://localhost:3000/api/v1/users/');
    const users = await response.json();
    return users;
  } catch (error) {
    console.log(error);
  }
};

export const fetchOneUser = async (id) => {
  try {
    const response = await fetch(`http://localhost:3000/api/v1/users/${id}`);
    const userInfo = await response.json();
    console.log(userInfo);
  } catch (error) {
    console.log(error);
  }
};

export const addUser = async (user) => {
  const response = await fetch('http://localhost:3000/api/v1/users', 
    {
      method: 'POST',
      body: JSON.stringify({
        userName: user.userName,
        email: user.email,
        password: user.password
      }),
      headers: {
        'Content-Type': 'application/json' 
      }
    });
  const userId = await response.json();
  return userId.id;
};

