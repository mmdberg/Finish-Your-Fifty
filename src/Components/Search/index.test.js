import Search from './index.js';
import { shallow } from 'enzyme';
import React from 'react';
import * as api from '../../apiCalls';
import { mockRace } from '../../mocks';

describe('Search', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Search />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should start with an empty state', () => {
    const expected = {
      state: '',
      results: [],
      error: ''
    };
    expect(wrapper.state()).toEqual(expected);
  });

  it('should update state on user input', () => {
    const mockEvent = {target:{name: 'state', value: 'C'}};
    const expected = {
      state: 'C',
      results: [],
      error: ''
    };

    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state()).toEqual(expected);
  });

  it('should call fetchRaces with the right params on submit', () => {
    const mockEvent = { preventDefault: jest.fn()};
    api.fetchRaces = jest.fn().mockImplementation(() => Promise.resolve(
      mockRace
    ));
    wrapper.setState({state: 'CA'});
    wrapper.instance().handleSubmit(mockEvent);
    expect(api.fetchRaces).toHaveBeenCalledWith('CA');
  });

  it('should set error message if there are no results', async () => {
    const mockEvent = { preventDefault: jest.fn() };
    const expected = 'No races match your search criteria. Try again.';
    api.fetchRaces = jest.fn().mockImplementation(() => Promise.resolve([]));
    await wrapper.instance().handleSubmit(mockEvent);
    expect(wrapper.state('error')).toEqual(expected);
  });

  it('should call displayResults with right params if results', async () => {
    const mockEvent = { preventDefault: jest.fn() };
    wrapper.instance().displayResults = jest.fn();
    api.fetchRaces = jest.fn().mockImplementation(() => Promise.resolve(
      [mockRace]
    ));
    await wrapper.instance().handleSubmit(mockEvent);
    expect(wrapper.instance().displayResults).toHaveBeenCalledWith([mockRace]);
  });

  it('should set state of results with jsx on submit after fetch', () => {
    const expected = 
      [<tr key='0'>
        <td>Spooky 5k</td>
        <td>San Diego</td>
        <td>04-05-2018</td>
        <td>Atlas Park</td>
      </tr>];
    wrapper.instance().displayResults([mockRace]);
    expect(wrapper.state('results')).toEqual(expected);
  });

  it('should reset the state', () => {
    const expected = {
      state: '',
      results: [],
      error: ''
    };
    wrapper.setState({
      state: 'FL',
      results: [mockRace]
    });
    wrapper.instance().clearSearch();
    expect(wrapper.state()).toEqual(expected);
  });


});