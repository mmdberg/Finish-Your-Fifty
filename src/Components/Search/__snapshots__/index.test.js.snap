// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Search />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Search for races by state
</h2>,
        <form>
          <input
                    maxLength="2"
                    name="state"
                    onChange={[Function]}
                    placeholder="State (ex. CA)"
                    type="text"
                    value=""
          />
          <input
                    maxLength="4"
                    name="year"
                    onChange={[Function]}
                    placeholder="Year"
                    type="text"
                    value=""
          />
          <button
                    onClick={[Function]}
          >
                    Search
          </button>
          <button
                    onClick={[Function]}
          >
                    Clear
          </button>
</form>,
        "",
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Search for races by state",
        },
        "ref": null,
        "rendered": "Search for races by state",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              maxLength="2"
              name="state"
              onChange={[Function]}
              placeholder="State (ex. CA)"
              type="text"
              value=""
/>,
            <input
              maxLength="4"
              name="year"
              onChange={[Function]}
              placeholder="Year"
              type="text"
              value=""
/>,
            <button
              onClick={[Function]}
>
              Search
</button>,
            <button
              onClick={[Function]}
>
              Clear
</button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "maxLength": "2",
              "name": "state",
              "onChange": [Function],
              "placeholder": "State (ex. CA)",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "maxLength": "4",
              "name": "year",
              "onChange": [Function],
              "placeholder": "Year",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Search",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Search",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Clear",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Clear",
            "type": "button",
          },
        ],
        "type": "form",
      },
      "",
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Search for races by state
</h2>,
          <form>
            <input
                        maxLength="2"
                        name="state"
                        onChange={[Function]}
                        placeholder="State (ex. CA)"
                        type="text"
                        value=""
            />
            <input
                        maxLength="4"
                        name="year"
                        onChange={[Function]}
                        placeholder="Year"
                        type="text"
                        value=""
            />
            <button
                        onClick={[Function]}
            >
                        Search
            </button>
            <button
                        onClick={[Function]}
            >
                        Clear
            </button>
</form>,
          "",
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Search for races by state",
          },
          "ref": null,
          "rendered": "Search for races by state",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                maxLength="2"
                name="state"
                onChange={[Function]}
                placeholder="State (ex. CA)"
                type="text"
                value=""
/>,
              <input
                maxLength="4"
                name="year"
                onChange={[Function]}
                placeholder="Year"
                type="text"
                value=""
/>,
              <button
                onClick={[Function]}
>
                Search
</button>,
              <button
                onClick={[Function]}
>
                Clear
</button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "maxLength": "2",
                "name": "state",
                "onChange": [Function],
                "placeholder": "State (ex. CA)",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "maxLength": "4",
                "name": "year",
                "onChange": [Function],
                "placeholder": "Year",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Search",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Search",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Clear",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Clear",
              "type": "button",
            },
          ],
          "type": "form",
        },
        "",
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
